native_extension/GoogleSignInANE/src/com/tuarua/google/GoogleSignIn.as

```diff
@@ -16,49 +16,95 @@
 
 package com.tuarua.google {
 import com.tuarua.fre.ANEError;
+import com.tuarua.google.signin.GoogleSignInOptions;
 
 import flash.events.EventDispatcher;
 
+[Event(type="com.tuarua.google.signin.events.GoogleSignInEvent", name="GoogleSignInEvent.SignIn")]
+[Event(type="com.tuarua.google.signin.events.GoogleSignInEvent", name="GoogleSignInEvent.Error")]
+
+/**
+ * Facade to GoogleSignIn api.
+ * You can use shared instance or create GoogleSignIn with constructor if additional options needed.
+ * 
+ * <listing version="3.0">
+ * function onSignIn(event:GoogleSignInEvent):void {
+ *     var account:GoogleSignInAccount = event.account;
+ *     trace('Hello, ' + account.displayName + '!');
+ *     var credential:GoogleAuthCredential = new GoogleAuthCredential(account.idToken, account.accessToken);
+ *     Firebase.auth().signInWithCredential(credential);
+ * }
+ *
+ * var api:GoogleSignIn = GoogleSignIn.shared()
+ * api.addEventListener(GoogleSignInEvent.SIGN_IN, onSignIn);
+ * api.signIn();
+ * </listing>
+ * 
+ * <listing version="3.0">
+ * // For sign in using Google Play Games you need to specify custom options
+ * 
+ * function onSignIn(event:GoogleSignInEvent):void {
+ *     var account:GoogleSignInAccount = event.account;
+ *     trace('Hello, ' + account.displayName + '!');
+ *     var credential:PlayGamesAuthCredential = new PlayGamesAuthCredential(account.serverAuthCode);
+ *     Firebase.auth().signInWithCredential(credential);
+ * }
+ * 
+ * var api:GoogleSignIn = new GoogleSignIn(new GoogleSignInOptions({
+ *     base: GoogleSignInOptions.DEFAULT_GAMES_SIGN_IN,
+ *     requestServerAuthCode: true
+ * }));
+ * 
+ * api.addEventListener(GoogleSignInEvent.SIGN_IN, onSignIn);
+ * api.signIn();
+ * </listing>
+ */
 public final class GoogleSignIn extends EventDispatcher {
     private static var _shared:GoogleSignIn;
-    private static var _apiKey:String;
+    private static var _serverClientId:String;
+    private var _context:GoogleSignInANEContext;
 
-    /** @private */
-    public function GoogleSignIn() {
-        if (GoogleSignInANEContext.context) {
-            var ret:* = GoogleSignInANEContext.context.call("init", _apiKey);
-            if (ret is ANEError) throw ret as ANEError;
+    public function GoogleSignIn(options:GoogleSignInOptions) {
```

Comment:
Although I appreciate the PR
I'm afraid I can't accept the complete unnecessary rewrite of both GoogleSignIn and GoogleSignInANEContext.
It's now both a singleton via `.shared()` and can be instantiated via `new()` ?
That makes no sense.

There was a simpler way of adding the options and which will conform to the rest of the codebase.

See how I have done for Firestore
https://github.com/tuarua/Firebase-ANE/blob/369e3c14c7a71d39ddc55532a0f804aaefc13d2c/native_extension/FirestoreANE/src/com/tuarua/firebase/Firestore.as#L29
https://github.com/tuarua/Firebase-ANE/blob/369e3c14c7a71d39ddc55532a0f804aaefc13d2c/native_extension/FirestoreANE/src/com/tuarua/firebase/Firestore.as#L117

Reply:
The main idea of having `.shared()` singleton was backward compatibility with older usages. Instantiation using `new` is preferable usage. That leaves additional options to a developer for having both Google Sign In and Play Games Sign In or using different clients (services) in one application.

That could be confusing because other ANEs are using the singleton pattern.

Reply:
Thanks for pointing me to `FirestoreSettings`. I will refer to that code.